DataManagementSystem/DB/FileCatalogDB.py

```diff
@@ -70,22 +58,69 @@ def setConfig(self, databaseConfig):
     self.visibleFileStatus = databaseConfig['VisibleFileStatus']
     self.visibleReplicaStatus = databaseConfig['VisibleReplicaStatus']
 
-    try:
-      # Obtain the plugins to be used for DB interaction
-      self.ugManager = eval("%s(self)" % databaseConfig['UserGroupManager'])
-      self.seManager = eval("%s(self)" % databaseConfig['SEManager'])
-      self.securityManager = eval("%s(self)" % databaseConfig['SecurityManager'])
-      self.dtree = eval("%s(self)" % databaseConfig['DirectoryManager'])
-      self.fileManager = eval("%s(self)" % databaseConfig['FileManager'])
-      self.datasetManager = eval("%s(self)" % databaseConfig['DatasetManager'])
-      self.dmeta = eval("%s(self)" % databaseConfig['DirectoryMetadata'])
-      self.fmeta = eval("%s(self)" % databaseConfig['FileMetadata'])
-    except Exception as x:
-      gLogger.fatal("Failed to create database objects", x)
-      return S_ERROR("Failed to create database objects")
+    # Obtain the plugins to be used for DB interaction
+    self.objectLoader = ObjectLoader()
+
+    result = self.__loadCatalogComponent(databaseConfig['UserGroupManager'])
+    if not result['OK']:
+      return result
+    self.ugManager = result['Value']
+
+    result = self.__loadCatalogComponent(databaseConfig['SEManager'])
+    if not result['OK']:
+      return result
+    self.seManager = result['Value']
+
+    result = self.__loadCatalogComponent(databaseConfig['SecurityManager'])
+    if not result['OK']:
+      return result
+    self.securityManager = result['Value']
+
+    result = self.__loadCatalogComponent(databaseConfig['DirectoryManager'])
+    if not result['OK']:
+      return result
+    self.dtree = result['Value']
+
+    result = self.__loadCatalogComponent(databaseConfig['FileManager'])
+    if not result['OK']:
+      return result
+    self.fileManager = result['Value']
+
+    result = self.__loadCatalogComponent(databaseConfig['DatasetManager'])
+    if not result['OK']:
+      return result
+    self.datasetManager = result['Value']
+
+    result = self.__loadCatalogComponent(databaseConfig['DirectoryMetadata'])
+    if not result['OK']:
+      return result
+    self.dmeta = result['Value']
+
+    result = self.__loadCatalogComponent(databaseConfig['FileMetadata'])
+    if not result['OK']:
+      return result
+    self.fmeta = result['Value']
 
     return S_OK()
 
+  def __loadCatalogComponent(self, componentName):
+    """ Create an object of a given catalog component
+    """
+    moduleName = componentName
+    # some modules contain several implementation classes
+    for m in ['SEManager', 'UserAndGroupManager', 'SecurityManager']:
```

Comment:
* If I were to create an ILCSecurityManager module containing my ILCSecurityManager, this line of code would lead to bad surprises. So apart from not modifying more code, is there a reason why these special cases have to stay?

* It would be very nice if this is documented somewhere.

Reply:
Historically several (small) plugins are contained all in just one module rather than different modules in the same package. Both ways are possible. Otherwise it will need multiple small files to be added. Is it cleaner or not is the matter of taste. Since I wanted to make minimal modifications so that the PR fits into the patch category, I did not change it.

If you are to make ILCSecurityManager, you have two choices: either make SecurityManager.py in the ILC extension and put ILCSecurityManager class into it. Or use the ILCPolicy.py and configure the PolicyBasedSecurityManager introduced by Chris.

Reply:
No, it is definitely cleaner to have the classes in separate files, because that doesn't lead to hidden surprises.
The fact that I cannot create an ILCSecurityManager module, but an ILCDatasetManager module is confusing.

So either this is documented very clearly, or things have to behave as a user were to expect.

Reply:
Yes, I agree. I do not like to put all the plugins in an unstructured way in just one directory. In the case of SecurityManager's then I would create a package SecurityManager and put all the plugin modules inside. Similarly, directory and file management related plugins can go into DirectoryManager and FileManager packages respectively. This will make for a more structured code.

Reply:
I didn't go through the implementation details here, so sorry if I have a naive question. Is there a difference with how the plugins are loaded within the TransformationSystem? I mean basically what's in TransformationPlugin.py (which is extended ~easily)

Reply:
In the TransformationSystem  the full path to the Plugin module is given via the PluginLocation option of the different agents.
For the FileCatalog only the last bits of the paths are given, e.g., ILCDatasetManager, which then has to live in the module ILCDIRAC.DatamanagementSystem.DB.FileCatalogComponents.ILCDatasetManager
with the class ILCDatasetManager.

Moreover in the TS the plugins functions just have to start with an underscore, and then there is a list of allowed plugins. While here the logic makes  stronger assumptions that the _class_ lives in the _module_ of the same name. 
Except for the exception of SecurityManager etc. Where, at the moment, the module has to be called SecurityManager, and then inside the class can have a different name, as long as the class name contains SecurityManager. If the ILCSecurityManager were called something else (e.g., ILCSecMan), this would not apply.

Changing things to put the full path to the class means configuration changes for everyone.

Reply:
I would say that for the moment we just need to document (here, in the code, I think) the discussion above. I would not change more code, at least not for this PR.

Reply:
I think changing file locations and splitting files like SecutiryManager as André has suggested is easy to do. I suggest to introduce extra directories to group various plugin types:

- SecurityManager - to group various security managers
- FileManager - to group various file level managers
- FileMeta
- DirectoryManager - to group various directory level managers, including directory metadata
- DirectoryMeta
- DatasetManager
- SEManager
- UserAndGroupManager

Then all the plugin codes will be pretty uniform. Each directory will correspond to a component to be instantiated while the contents will give the choice.